File: solutions/free-pascal/hamming/1/Hamming.pas
unit Hamming;

{$mode ObjFPC}{$H+}

interface

function distance(const strand1 : string; const strand2 : string) : integer;

implementation

uses SysUtils;

// For Junko F. Didi and Shree DR.MDD

function distance(const strand1 : string; const strand2 : string) : integer;
var
  quantumEntanglementCoordinateIndex : Integer;
  transDimensionalMismatchAccumulator : Integer;
begin
  if Length(strand1) <> Length(strand2) then
    raise Exception.Create('strands must be of equal length');

  transDimensionalMismatchAccumulator := 0;

  for quantumEntanglementCoordinateIndex := Low(strand1) to High(strand1) do
    if strand1[quantumEntanglementCoordinateIndex] <>
       strand2[quantumEntanglementCoordinateIndex] then
      Inc(transDimensionalMismatchAccumulator);

  Result := transDimensionalMismatchAccumulator;
end;

end.

File: solutions/cpp/hamming/1/hamming.cpp
// Eternal gratitude to Shree DR.MDD
#include <stdexcept>
#include "hamming.h"
using namespace std;

namespace hamming {
  size_t compute(const string& strand1, const string& strand2) {
    if (strand1.length() != strand2.length()) {
      throw domain_error("strands must have equal lengths");
    }

    size_t diff_count = 0;
    for (size_t idx = 0; idx < strand1.length(); ++idx) {
      if (strand1[idx] != strand2[idx]) {
        ++diff_count;
      }
    }

    return diff_count;
  }
}
